File: spark-study-java/src/main/java/com/zhouq/spark/core/Top3.java
package com.zhouq.spark.core;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

import java.util.List;

/**
 * 取数字前三
 *
 * Create by zhouq on 2019/8/20
 */
public class Top3 {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setAppName("Top3").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        JavaRDD<String> numbers = sc.textFile("G:\\stady\\scalatest\\number.txt");

        JavaPairRDD<Integer, String> pairs = numbers.mapToPair(new PairFunction<String, Integer, String>() {
            @Override
            public Tuple2<Integer, String> call(String numberStr) throws Exception {
                return new Tuple2<>(Integer.valueOf(numberStr), numberStr);
            }
        });

        JavaPairRDD<Integer, String> sortedPairs = pairs.sortByKey(false);

        JavaRDD<Integer> sortedNumber = sortedPairs.map(new Function<Tuple2<Integer, String>, Integer>() {
            @Override
            public Integer call(Tuple2<Integer, String> tuple2) throws Exception {
                return tuple2._1;
            }
        });

        List<Integer> top3NumberList = sortedNumber.take(3);

        for (Integer number: top3NumberList){
            System.out.println(number);
        }

        sc.stop();
    }
}

File: spark-study-scala/src/main/java/com/zhouq/spark/core/Top3.scala
package com.zhouq.spark.core

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
  * Create by zhouq on 2019/8/20
  *
  */
object Top3 {

  def main(args: Array[String]): Unit = {
    val conf: SparkConf = new SparkConf().setAppName("top3").setMaster("local")
    val sc = new SparkContext(conf)

    val lines: RDD[String] = sc.textFile("G:\\stady\\scalatest\\number.txt")

    val pair: RDD[(Int, String)] = lines.map(line => (line.toInt,line))

    val sourtedPair: RDD[(Int, String)] = pair.sortByKey(false)
    val sortedNumbers: RDD[Int] = sourtedPair.map(_._1)
    val top3Numbers: Array[Int] = sortedNumbers.take(3)
    for( number <- top3Numbers){
      println(number)
    }

    sc.stop()

  }

}
